examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
+    {
+      return !(a == b);
+    }
+    public override bool Equals(Object other)
+    {
+      if (ReferenceEquals(null, other)) return false;
+      if (ReferenceEquals(this, other)) return true;
+      if (other.GetType() != GetType()) return false;
+      return this.Equals(other as MapToInside_Key);
+    }
+    public bool Equals(MapToInside_Key other)
+    {
+      return !ReferenceEquals(null, other)
+        && SourceAddress.Equals(other.SourceAddress)
+        && SourcePort.Equals(other.SourcePort)
+        && ArrivalPort.Equals(other.ArrivalPort);
     }
+    public override int GetHashCode() { return hashCode; }
+    public override string ToString()
+    {
+#if DEBUG
+      if (asString != null)
+        return asString;
+      else
+#endif
+      return SourceAddress.ToString() + ":" + SourcePort.ToString() + " to :" + ArrivalPort.ToString();
+    }
+  }
+  sealed class MapToOutside_Key : IEquatable<MapToOutside_Key>
+  {
+    // FIXME we should really use immutable values for IPAddress since we are using a dictionary
+    private readonly IPAddress _SourceAddress, _DestinationAddress;
+    private readonly ushort _SourcePort, _DestinationPort;
+    public IPAddress SourceAddress { get { return _SourceAddress; } }
+    public ushort SourcePort { get { return _SourcePort; } }
+    public IPAddress DestinationAddress { get { return _DestinationAddress; } }
+    public ushort DestinationPort { get { return _DestinationPort; } }
+    private readonly int hashCode;
+#if DEBUG
+    private string asString;
+#endif
 
-    public override int GetHashCode() {
-      // FIXME this is a bit heavy to compute
-      return this.ToString().GetHashCode();
+    public MapToOutside_Key(IPAddress sourceAddress, ushort sourcePort,
+                            IPAddress destinationAddress, ushort destinationPort)
+    {
+      if (ReferenceEquals(null, sourceAddress)) throw new ArgumentNullException("sourceAddress");
+      if (ReferenceEquals(null, destinationAddress)) throw new ArgumentNullException("destinationAddress");
+
+      _SourceAddress = sourceAddress;
+      _SourcePort = sourcePort;
+      _DestinationAddress = destinationAddress;
+      _DestinationPort = destinationPort;
+      // FIXME computing hash at construction relies on address being immutable (addr.Scope can change)
+      hashCode = new { SourceAddress, SourcePort, DestinationAddress, DestinationPort }.GetHashCode();
+#if DEBUG
+      asString = this.ToString();
+#endif
     }
 
-    public override string ToString() {
-      return ("(" +
-          ip_address.ToString() +
-          ":" + tcp_port.ToString() +
-          " " + (assigned_tcp_port.HasValue ? Convert.ToString(assigned_tcp_port.Value) : ".") +
-          "|" + (network_port.HasValue ? Convert.ToString(network_port.Value) : ".") +
-          ")");
+    public static bool operator ==(MapToOutside_Key a, MapToOutside_Key b)
+    {
+      return !ReferenceEquals(null, a) && a.Equals(b);
```

Comment:
Makes sense, thanks. Feel free to prune the code, and leave a comment to 
say that, for example, it was sufficient to override Equals and not the 
operators.

On 2016-07-07 13:56, TMVector wrote:
> In examples/NAT.cs [1]:
> 
>> }
>> 
>> - public override string ToString() {
>> - return ("(" +
>> - ip_address.ToString() +
>> - ":" + tcp_port.ToString() +
>> - " " + (assigned_tcp_port.HasValue ?
> Convert.ToString(assigned_tcp_port.Value) : ".") +
>> - "|" + (network_port.HasValue ?
> Convert.ToString(network_port.Value) : ".") +
>> - ")");
>> + public static bool operator ==(MapToOutside_Key a,
> MapToOutside_Key b)
>> + {
>> + return !ReferenceEquals(null, a) && a.Equals(b);
> 
> Yes, but because Equals(MapToOutside_Key) is an instance method, we
> need to check for null before we can call it.
> 
> Since we only implement equality to allow use in the dictionaries, we
> don't really need to overload the operators. In my latest refactoring
> (not pushed and on a different branch), I have removed them to
> simplify the code. What do you think?
> 
> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/3#discussion_r69900767
> [2]
> https://github.com/niksu/pax/pull/3/files/d721adebf668133ae20c55f466ccfaa2344f51a2#r69900767
> [3]
> https://github.com/notifications/unsubscribe/ABDTTvpVBMqydTCv6ZHvg_fnwmrQ6hDhks5qTPeagaJpZM4JFDoJ

-- 
http://www.cl.cam.ac.uk/~ns441/